src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs

```diff
@@ -1580,7 +1900,376 @@ static void M(dynamic d1, dynamic d2)
+  IL_0002:  newobj     ""C..ctor()""
+  IL_0007:  stloc.1
+  IL_0008:  ldloc.0
+  IL_0009:  brtrue.s   IL_000d
+  IL_000b:  ldloc.1
+  IL_000c:  stloc.0
+  IL_000d:  ldloc.0
+  IL_000e:  callvirt   ""int object.GetHashCode()""
+  IL_0013:  call       ""void System.Console.WriteLine(int)""
+  IL_0018:  ret
+}
+");
+        }
+
+        [Fact]
+        public void Associativity()
+        {
+            var source = @"
+using System;
+class C
+{
+    public static void Main()
+    {
+        int? a1 = null, b1 = null, c1 = 1;
+        Console.WriteLine(a1 ??= b1 ??= c1);
+        int? a2 = null, b2 = null, c2 = 1;
+        Console.WriteLine(a2 ??= (b2 ??= c2));
+    }
+}";
+
+            CompileAndVerify(source, expectedOutput: @"
+1
+1
+");
+        }
+
+        [Fact]
+        public void ThisDisallowed()
+        {
+            var source = @"
+class C
+{
+    public void M1()
+    {
+        this ??= new C();
+    }
+}
+struct S
+{
+    public void M2()
+    {
+        this ??= new S();
+    }
+}
+";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(new DiagnosticDescription[] {
+                // (6,9): error CS1604: Cannot assign to 'this' because it is read-only
+                //         this ??= new C();
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "this").WithArguments("this").WithLocation(6, 9),
+                // (13,9): error CS0019: Operator '??=' cannot be applied to operands of type 'S' and 'S'
+                //         this ??= new S();
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "this ??= new S()").WithArguments("??=", "S", "S").WithLocation(13, 9)
+            });
+        }
+
+        [Fact]
+        public void DefiniteAssignment()
+        {
+            var source = @"
+class C
+{
+    void M()
+    {
+        object o1, o2;
+        C c1 = new C();
+
+        GetRef(o1 = """") ??= GetObject(o1);
+        o1.ToString();
+
+
+        GetRef() ??= GetObject(o2 = """", o2);
+        o2.ToString();
+
+        GetRef() ??= OutVar(out var o3).GetObject(o3);
+        o3.ToString();
+    }
+
+    object o;
+    ref object GetRef(object param = null) => ref o;
+    object GetObject(object param1, object param2 = null) => null;
+    C OutVar(out object o3)
+    {
+        o3 = null;
+        return this;
+    }
+}";
+
+            CreateCompilation(source).VerifyDiagnostics(new DiagnosticDescription[] {
+                // (14,9): error CS0165: Use of unassigned local variable 'o2'
+                //         o2.ToString();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "o2").WithArguments("o2").WithLocation(14, 9),
+                // (17,9): error CS0165: Use of unassigned local variable 'o3'
+                //         o3.ToString();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "o3").WithArguments("o3").WithLocation(17, 9)
+            });
+        }
+
+        [Fact]
+        public void COMIndexedProperties()
```

Comment:
Why are COM indexed properties different than normal indexed properties? Sure there is a reason but interested to know.

Reply:
I'm not entirely sure, @cston? They were on the test plan so I added a test for them.

---
In reply to: [215776649](https://github.com/dotnet/roslyn/pull/29691#discussion_r215776649) [](ancestors = 215776649)

Reply:
There may not be a meaningful difference since both indexers and indexed properties are bound as `BoundIndexerAccess`, but it seemed like an easy test case to include.

---
In reply to: [216046319](https://github.com/dotnet/roslyn/pull/29691#discussion_r216046319) [](ancestors = 216046319,215776649)

Reply:
@jaredpar unless you have an objection, I'm going to leave this in.

---
In reply to: [216049568](https://github.com/dotnet/roslyn/pull/29691#discussion_r216049568) [](ancestors = 216049568,216046319,215776649)